### PR TITLE
JP-1962: Data models for new NIRISS SOSS ref files

### LINKED ISSUE #5840: Create data models for new NIRISS SOSS extraction reference files
_Issue [JP-1962](https://jira.stsci.edu/browse/JP-1962) was created on JIRA by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse):_

The NIRISS IDT has delivered new reference files to be used as part of the enhanced extract_1d methods they are developing for SOSS mode spectra. The new reference files need to have data models created for them, in order to allow delivery to CRDS and for use in the cal pipeline.

### PR DESCRIPTION
**Pull request by hbushouse**
Closes #5840 
Resolves [JP-1962](https://jira.stsci.edu/browse/JP-1962)

**Description**

This PR implements data models for the 4 new types of reference files that've been created by the NIRISS IDT for use in their optimal SOSS 1D spectral extraction algorithms.

Checklist
- [ ] Tests
- [ ] Documentation
- [x] Change log
- [X] Milestone
- [X] Label(s)

### CODE REVIEW

**Comment by hbushouse** [comment]
Reviewers: 3 of these new models (SpecProfile, SpecTrace, and WaveMap) are for reference files that contain multiple extensions that use the same EXTNAME (but unique EXTVER's) so that we need data models that mimic models like `MultiSlitModel`. Instead of using bulky names like `MultiSpecProfileModel` for the top-level classes I decided to go with the simpler and more straightforward names `SpecProfileModel`, `WaveMapModel`, etc. Then to make the lower-level model names unique I've called them `SpecProfileSingleModel`, `WaveMapSingleModel`, etc. For the purpose of just loading and using these ref files in the pipeline (which will always need the "multi" form) no one will ever see or use the lower-level `xxxSingleModel` names.

If anyone has a better way to organize or implement these crazy hierarchical model structures, I'm very open to suggestions.

Testing: I've tested all of them by using them to load the FITS ref files that've already been created and they work fine, allowing access to all necessary meta data and data arrays.

Documentation: I'm going to wait until the new NIRISS SOSS extract_1d step code is delivered to update the reference file docs to list these new ref file models.

**Comment by hbushouse** [comment]
Failure in CI Latest Dependencies test seems to be unrelated (a problem in date-time conversions/calculations)

**Comment by stscieisenhamer** [comment]
This may be too subtle, but using the English plural noun rules would be more concise:

- specprofile.schema/SpecProfileModel (single) => specprofiles.schema/SpecProfilesModel (multi)
- spectrace.schema/SpecTraceModel => spectraces.schema/SpectTracesModel
- wavemap.schema/WaveMap => wavemaps.schema/WaveMapsModel

**Review by stscieisenhamer — APPROVED** [review]

**Comment by hbushouse** [comment]
> This may be too subtle, but using the English plural noun rules would be more concise:
> 
> * specprofile.schema/SpecProfileModel (single) => specprofiles.schema/SpecProfilesModel (multi)
> * spectrace.schema/SpecTraceModel => spectraces.schema/SpectTracesModel
> * wavemap.schema/WaveMap => wavemaps.schema/WaveMapsModel

Ooh, I hadn't thought of that possibility. But yes, you would need to be very careful to make sure to distinguish the plural vs singular versions in use.

File: jwst/datamodels/schemas/specprofilesingle.schema.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+- $ref: subarray.schema
+- type: object
+  properties:
+    data:
+      title: The profile data
+      fits_hdu: ORDER
+      max_ndim: 2
+      datatype: float32
+- type: object
+  properties:
```
Comment (nden) [inline jwst/datamodels/schemas/specprofilesingle.schema.yaml:18]: If the following three are going directly under `tree` and not under `meta` there's no need of another `object` definition.
And the same for the rest of the files.

File: jwst/datamodels/specprofile.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    """
+    schema_url = "http://stsci.edu/schemas/jwst_datamodel/specprofile.schema"
+
+    def __init__(self, init=None, **kwargs):
+
+        if isinstance(init, SpecProfileSingleModel):
+            super(SpecProfileModel, self).__init__(init=None, **kwargs)
+            self.profile.append(self.profile.item())
+            self.profile[0].data = init.data
+            return
```
Comment (nden) [inline jwst/datamodels/specprofile.py:40]: Why is this return here, in an `__init__` method?

File: jwst/datamodels/schemas/speckernel.schema.yaml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+          title: Wavelength of the last kernel (in microns)
+          type: number
+          fits_keyword: WAVEMAX
+          fits_hdu: WAVELENGTHS
+        specos:
+          title: The spectral oversampling of the native pixels
+          type: integer
+          fits_keyword: SPECOS
+          fits_hdu: KERNELS
+        halfwidt:
```
[the remark this answers is not in the document]
Comment (hbushouse) [reply jwst/datamodels/schemas/speckernel.schema.yaml:33]: That's the way the FITS keyword name is spelled, in order to keep it to 8 chars. I made all the datamodels attribute names the same as the keyword names. I suppose we could spell them out more completely.

File: jwst/datamodels/schemas/speckernel.schema.yaml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+          title: The half width of the kernel in native pixels
+          type: integer
+          fits_keyword: HALFWIDT
+          fits_hdu: KERNELS
+        indcentr:
```
[the remark this answers is not in the document]
Comment (hbushouse) [reply jwst/datamodels/schemas/speckernel.schema.yaml:38]: Yes. We could make the attribute names more readable.

File: jwst/datamodels/schemas/specprofilesingle.schema.yaml (hunk 1 above)
Comment (hbushouse) [reply jwst/datamodels/schemas/specprofilesingle.schema.yaml:18]: Oh, OK. I'll try making those changes.

File: jwst/datamodels/schemas/speckernel.schema.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (hbushouse) [reply jwst/datamodels/schemas/speckernel.schema.yaml]: Updated to spell out the name in the latest commit

File: jwst/datamodels/schemas/speckernel.schema.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (hbushouse) [reply jwst/datamodels/schemas/speckernel.schema.yaml]: Updated in latest commit

File: jwst/datamodels/schemas/specprofilesingle.schema.yaml (hunk 1 above)
Comment (hbushouse) [reply jwst/datamodels/schemas/specprofilesingle.schema.yaml]: Done

File: jwst/datamodels/specprofile.py (hunk 2 above)
Comment (hbushouse) [reply jwst/datamodels/specprofile.py]: Removed all returns

**Pull request merged**